src/Networking/NexusMods.Networking.ModUpdates/PerFeedCacheUpdater.cs

```diff
@@ -0,0 +1,155 @@
+using System.Diagnostics;
+using NexusMods.Networking.ModUpdates.Private;
+using NexusMods.Networking.ModUpdates.Structures;
+using NexusMods.Networking.ModUpdates.Traits;
+namespace NexusMods.Networking.ModUpdates;
+
+/// <summary>
+/// This is a helper struct which helps us make use of the 'most recently updated mods for game'
+/// API endpoint to use as a cache. (Where 'game' is a 'feed')
+/// For full info on internals, see the NexusMods.App project documentation.
+///
+/// This API consists of the following:
+///
+/// 1. Input [Constructor]: A set of items with a 'last update time' (see <see cref="ICanGetLastUpdatedTimestamp"/>)
+///           and a 'unique id' (see <see cref="ICanGetUid"/>) that are relevant to the current 'feed' (game).
+/// 
+/// 2. Update [Method]: Submit results from API endpoint returning 'most recently updated mods for game'.
+///           This updates the internal state of the <see cref="MultiFeedCacheUpdater{TUpdateableItem}"/>.
+/// 
+/// 3. Output [Info]: The <see cref="MultiFeedCacheUpdater{TUpdateableItem}"/> outputs items with 2 categories:
+///        - Up-to-date mods. These should have their timestamp updated.
+///        - Out of date mods. These require re-querying the data from external source.
+/// </summary>
+/// <remarks>
+/// Within the Nexus Mods App:
+///
+/// - 'Input' is our set of locally cached mod pages.
+/// - 'Update' is our results of `updated.json` for a given game domain.
+/// - 'Output' are the pages we need to update.
+///
+/// The 'Feed' in the context of the Nexus App is the individual game's 'updated.json' endpoint;
+/// i.e. a 'Game Mod Feed'
+/// </remarks>
+public class PerFeedCacheUpdater<TUpdateableItem> where TUpdateableItem : ICanGetLastUpdatedTimestamp, ICanGetUid
+{
+    private readonly TUpdateableItem[] _items;
+    private readonly Dictionary<ModId, int> _itemToIndex;
+    private readonly CacheUpdaterAction[] _actions;
+
+    /// <summary>
+    /// Creates a (per-feed) cache updater from a given list of items for which
+    /// we want to check for updates.
+    /// </summary>
+    /// <param name="items">The items to check for updates.</param>
+    /// <param name="expiry">
+    ///     Maximum age before an item has to be re-checked for updates.
+    ///     The max for Nexus Mods API is 1 month.
+    /// </param>
+    /// <remarks>
+    ///     In order to ensure accuracy, the age field should include the lifespan of
+    ///     the <see cref="PerFeedCacheUpdater{TUpdateableItem}"/> as well as the cache
+    ///     time on the server's end. That should prevent any technical possible
+    ///     eventual consistency errors due to race conditions. Although
+    /// </remarks>
+    public PerFeedCacheUpdater(TUpdateableItem[] items, TimeSpan expiry)
+    {
```

Comment:
If we go off of MnemonicDB cache items, we can replace this class with a query over items that are older than a certain value. No reason to keep all this in-memory

Reply:
Can you link me to the relevant API for this? I'm curious actually.

Reply:
We already have `NexusModsFileMetadata` and Mod metadata, we can extend those models to include the last updated date. then do a `db.Datoms(NexusModPageMeatadata.LastUpdated)` and do a where on that to check for those that are older than a given date

Reply:
Ah, the relevant research part for that is `1. Determine Updated Mod Pages`.
Let's break down the problem (concisely).

> [!NOTE]  
> V2 API fetches max 50 items at once.

# Core Issue

1. User has a lot of mods
2. Lot of mods == lot of API calls
3. We should reduce API calls.

## Proposed Alternative

Filter out (exclude) items more recent than a set TimeSpan, and check only the mod pages older than that.

### Effect

This throttles the update check rate; leading to a reduction of API calls.
Reduction is proportional to throttle rate.

### Cache Hit Rate

> [!NOTE]  
> Assuming 300 app launches in a month, i.e. 10 per day.
> And 1000 mods.

At 4 week throttle: (1000 / 50) API Calls
At 2 weeks throttle: (2000 / 50) API Calls
At 1 week throttle: (4000 / 50) API Calls

### Side Effects

Updates for mods get delayed for a while.

Is this a problem?

Normally no, BUT if a game gets updated, breaking mods, the user may require to see mod updates in a timely manner.

This is a concern for big games, for example Baldur's Gate 3. If a user obtains the game from a store that does not keep older versions (e.g. Epic, Xbox) and don't do a full game backup, they're cooked.

Alternatively if a mod author or server informs users of an update, the user should not be expected to use a 'fetch updates manually' button, which isn't part of the designs either.

## Current Solution

Obtain a list of last updated timestamps for installed mods via the API.
Process results via rules described in `1. Determine Updated Mod Pages`.

### Effect

- Updates can be checked on a more regular basis (e.g. every app boot).
    - Time to detected update may be 1 hour (if we refresh every hour), as opposed to 1/2/4 weeks.

- Requires a batch timestamp API call to backend. (Available in V1, not yet in V2)

- Switchover from V1 to V2 API is cheap (i.e. 1 hour to define the query, make wrapper struct, test code).

- Website can cache query for updated items; making computation cheap and cache hit rates high.
    - Currently V1's API caches for 5 minutes before recomputing.
    - Because everyone does same query, nothing needs recomputed on per-game basis until 5 minutes have elapsed.

### Side Effects

Currently only has V1 API. V2 API is a request (ticket) to backend.
However, with this implementation, constraints (requirements) for ticket to backend are well defined.

### Cache Hit Rate

> [!NOTE]  
> Assuming 300 app launches in a month, i.e. 10 per day.
> And 1000 mods.

***Initial Fetch***:

- (1000 / 50) API calls *if never checked before* or *all items are expired* (user hasn't launched App in 1 month).

***Subsequent Fetches***:

(Where `$NumGames` is the number of game feeds we downloaded mods from)
(Where `$NumToUpdate` is number of mod pages updated since we last checked)

- `$NumGames` API calls (typically 1 or 2) if last update check was ran less than `expiry` (currently 1 month).
- (`$NumToUpdate` / 50) API calls (typically 0/1) for any mods which have been updated since last check.

Assuming an extreme unrealistic scenario where we check for updates every hour, and the user has the App open for 12 hours a day, this means 336 API calls over a month. (+1 for Number Of Times any Mod Received an Update)

-------------

[Apologies for late response, was preoccupied, then quiz night happened]

Reply:
The issue I have with all of this is that it's overkill for the current ticket. I would like to have a conversation with the backend team, and others before deciding on the proper way to pull and query updates. 

For now all we need is:

1) a way to "check for updates" we can make this a button, or a single API call, with the number of users/mods we currently work with this isn't an issue. 
2) as mentioned in the initial kickoff meeting it's completely fine to do this check at startup as we're still in the early development phase
3) we can then write a mnemonicDB query that finds possible update candiates 
4) we can display those updates to the user in the UI

The nice thing about this approach is that it segments out the different parts. We already store mod metadata in MDB, we already query it when building our datagrids, so the whole project breaks down into 3 clean parts:

1) looking for updates in MDB
2) performing an update given two `NexusLibraryFile` instances
3) updating the cached data in MDB using some critera

We don't need it to be generic, split off, or even worry about the number of API calls at this point. Once we get the above built it will be simple enough to talk to the backend and get a implementation for whatever update query logic we would like to have.